### PR TITLE
rgw: Introduce BucketEncryption APIs to support SSE-S3 feature

### PR DESCRIPTION
**Pull request by rahuldevp**
This patch introduces support for 3 new BucketEncryption APIs which are listed
below and are helpful in supporting AWS SSE-S3 encryption mode.
PutBucketEncryption: https://docs.aws.amazon.com/AmazonS3/latest/API/API_PutBucketEncryption.html
GetBucketEncryption: https://docs.aws.amazon.com/AmazonS3/latest/API/API_GetBucketEncryption.html
DeleteBucketEncryption: https://docs.aws.amazon.com/AmazonS3/latest/API/API_DeleteBucketEncryption.html

The user provided parameters are parsed and stored in the bucket's extended
attributes RGW_ATTR_BUCKET_ENCRYPTION and
RGW_ATTR_BUCKET_ENCRYPTION_SSE_S3_KEY_ID.

Signed-off-by: Rahul Dev Parashar rahul.dev@flipkart.com

## Checklist
- [ ] References tracker ticket
- [x] Updates documentation if necessary
- [x] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/rgw/rgw_common.h [hunk 1]
```diff
@@ -143,6 +144,10 @@ using ceph::crypto::MD5;
 #define RGW_ATTR_CRYPT_CONTEXT  RGW_ATTR_CRYPT_PREFIX "context"
 #define RGW_ATTR_CRYPT_DATAKEY  RGW_ATTR_CRYPT_PREFIX "datakey"
 
+/* SSE-S3 Encryption Attributes */
+#define RGW_ATTR_BUCKET_ENCRYPTION RGW_ATTR_PREFIX "encryption."
+#define RGW_ATTR_BUCKET_ENCRYPTION_SSE_S3_KEY_ID RGW_ATTR_BUCKET_ENCRYPTION "sse-s3-key-id"
```
Comment (cbodley) [inline src/rgw/rgw_common.h:149]: the trailing `.` seems odd for the `"encryption."` policy attribute

this looks similar to `#define RGW_ATTR_CRYPT_PREFIX   RGW_ATTR_PREFIX "crypt."` above, but we don't use `RGW_ATTR_CRYPT_PREFIX` directly as an attribute name

how about something like this?
```c++
#define RGW_ATTR_BUCKET_ENCRYPTION_PREFIX RGW_ATTR_PREFIX "sse-s3."
#define RGW_ATTR_BUCKET_ENCRYPTION_POLICY RGW_ATTR_BUCKET_ENCRYPTION_PREFIX "policy"
#define RGW_ATTR_BUCKET_ENCRYPTION_KEY_ID RGW_ATTR_BUCKET_ENCRYPTION_PREFIX "key-id"
```

File: src/rgw/rgw_common.h [hunk 2]
```diff
@@ -39,6 +39,7 @@
 #include "cls/rgw/cls_rgw_types.h"
 #include "include/rados/librados.hpp"
 #include "rgw_public_access.h"
+#include "rgw_bucket_encryption.h"
```
Comment (cbodley) [inline src/rgw/rgw_common.h:42]: it doesn't look like `rgw_common.h` depends on anything from this header

**Comment by mattbenjamin** [comment]
this is great progress, @rahuldevp

File: src/rgw/rgw_common.h (hunk 1 above)
Comment (rahuldevp) [reply src/rgw/rgw_common.h:149]: This looks better Casey. I will make the necessary changes along with other suggestions.

**Comment by rahuldevp** [comment]
> this is great progress, @rahuldevp

Thanks @mattbenjamin

File: src/rgw/rgw_op.cc [hunk 3]
```diff
@@ -8510,3 +8510,126 @@ void RGWDeleteBucketPublicAccessBlock::execute(optional_yield y)
...
+{
+  if (!verify_bucket_permission(this, s, rgw::IAM::s3GetBucketEncryption)) {
+    return -EACCES;
+  }
+  return 0;
+}
+
+void RGWGetBucketEncryption::execute(optional_yield y)
+{
+  auto attrs = s->bucket_attrs;
```
[the remark this answers is not in the document]
Comment (rahuldevp) [reply src/rgw/rgw_op.cc:8585]: @cbodley Need a little more clarification about this change Casey. Do you mean that I should replace 
`auto attrs = s->bucket_attrs;
  if (auto aiter = attrs.find(RGW_ATTR_BUCKET_ENCRYPTION);
      aiter == attrs.end()) {`
with
`auto aiter = s->bucket_attrs.find(RGW_ATTR_BUCKET_ENCRYPTION_POLICY);
  if (aiter == s->bucket_attrs.end()) {`
or any other suggestion is also there?

[the remark this answers is not in the document]
Comment (rahuldevp) [reply src/rgw/rgw_op.cc]: @cbodley Can you please confirm on this if my above comment is the right way or you were suggesting something else?

[the remark this answers is not in the document]
Comment (cbodley) [reply src/rgw/rgw_op.cc]: the only part that needs to change is this:
```diff
-  auto attrs = s->bucket_attrs;
+  const auto& attrs = s->bucket_attrs;
```

File: src/rgw/rgw_kms.cc [hunk 4]
```diff
@@ -159,16 +159,35 @@ add_name_val_to_obj(const char *n, std::string &v, rapidjson::GenericValue<E,A>
...
+class SSEContext {
+protected:
+  virtual ~SSEContext(){};
+public:
+  virtual std::string & backend() = 0;
+  virtual std::string & addr() = 0;
+  virtual std::string & auth() = 0;
+  virtual std::string & k_namespace() = 0;
+  virtual std::string & prefix() = 0;
+  virtual const std::string & secret_engine() = 0;
```
[the remark this answers is not in the document]
Comment (rahuldevp) [reply src/rgw/rgw_kms.cc]: This change was pushed for implementing SSE-S3 but later reverted to be raised as a separate PR. So resolving this conversation as it is no longer relevant here.

File: src/rgw/rgw_bucket_encryption.cc [hunk 5]
```diff
@@ -0,0 +1,41 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab ft=cpp
+//
+#include "rgw_bucket_encryption.h"
+#include "rgw_xml.h"
+
+void ApplyServerSideEncryptionByDefault::decode_xml(XMLObj *obj) {
+  RGWXMLDecoder::decode_xml("KMSMasterKeyID", kmsMasterKeyID, obj, false);
+  if(kmsMasterKeyID.compare("") != 0) {
+    throw RGWXMLDecoder::err("implementation for KMS is not supported yet");
```
[the remark this answers is not in the document]
Comment (rahuldevp) [reply src/rgw/rgw_bucket_encryption.cc]: Moved checks to `RGWPutBucketEncryption::execute()` and specific error codes are returned. For "aws:kms" and algorithm other than AES256 we are returning `ERR_NOT_IMPLEMENTED` with the respective message.

File: src/rgw/rgw_bucket_encryption.h [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+public:
+  ApplyServerSideEncryptionByDefault(): kmsMasterKeyID(""), sseAlgorithm("") {};
+
+  string kms_master_key_id() const {
+    return kmsMasterKeyID;
+  }
+
+  string sse_algorithm() const {
+    return sseAlgorithm;
+  }
```
[the remark this answers is not in the document]
Comment (rahuldevp) [reply src/rgw/rgw_bucket_encryption.h]: Done

**Comment by rahuldevp** [comment]
@cbodley I have addressed review comments and fixed checks which were failing. I had to squash and merge old commits as an old commit was failing Signed-off-by check.

**Comment by cbodley** [comment]
thanks @rahuldevp, could you please add some simple test cases of these APIs to https://github.com/ceph/s3-tests/?

**Comment by prasakri** [comment]
> thanks @rahuldevp, could you please add some simple test cases of these APIs to https://github.com/ceph/s3-tests/?

https://github.com/ceph/s3-tests/pull/404

**Review by cbodley — APPROVED** [review]

**Comment by cbodley** [comment]
👍 https://pulpito.ceph.com/cbodley-2021-07-26_21:18:53-rgw-wip-cbodley-testing-distro-basic-smithi/

> 2021-07-26T22:46:01.694 INFO:teuthology.orchestra.run.smithi101.stderr:s3tests_boto3.functional.test_s3.test_put_bucket_encryption ... ok
2021-07-26T22:46:02.619 INFO:teuthology.orchestra.run.smithi101.stderr:s3tests_boto3.functional.test_s3.test_get_bucket_encryption ... ok
2021-07-26T22:46:03.552 INFO:teuthology.orchestra.run.smithi101.stderr:s3tests_boto3.functional.test_s3.test_delete_bucket_encryption ... ok

**Pull request merged**